api/lib/flight_job_script_api/subprocess.rb

```diff
@@ -0,0 +1,191 @@
+#==============================================================================
+# Copyright (C) 2021-present Alces Flight Ltd.
+#
+# This file is part of Flight Job Script Service.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# Flight Job Script Service is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Job Script Service. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Job Script Service, please visit:
+# https://github.com/openflighthpc/flight-job-script-service
+#==============================================================================
+
+module FlightJobScriptAPI
+  class Subprocess
+    class Result < Struct.new(:stdout, :stderr, :exitstatus, :pid); end
+
+    def initialize(dir:, env:, gid:, logger:, timeout:, uid:)
+      @dir = dir
+      @env = env
+      @gid = gid
+      @logger = logger
+      @timeout = timeout
+      @uid = uid
+    end
+
+    def run(cmd, stdin, &block)
```

Comment:
I'm not sure why `cmd`/`block` are defined on `run` and `env`/`dir` on `initialize`. All four variables are used to start the command, so it would make sense to defined them in one place. Either on `initialize` or `run`.

Reply:
What about

```ruby
sp = SubProcess.new(env: env, dir: dir, **opts)
scripts = sp.run('list-scripts`)
prime_scripts_cache(scripts)
jobs = sp.run('list-jobs')
prime_jobs_cache(jobs)
jobs
```

Reply:
Is there an advantage to trying to reuse the same `SubProcess` class?

In most server applications, it will be re-created on each request. Also some requests will need to modify the `env`?

It maybe better to recreate the command each time within `initialize` and the call `run` to preform the heavy lifting? Any of the following:

```
# Two-step method
SubProcess.new(env, *cmd, **options).tap(&:run)

# One-step method
SubProcess.run(env, *cmd, **options)

# NOTE: env should go first to be consistent with `spawn` 
```

Reply:
> Is there an advantage to trying to reuse the same SubProcess class?

Probably not, but it seems "right".  A class method `run` for the common case seems like a nice addition at some point.